venom/rpc/reflect/openapi.py

```diff
@@ -1,6 +1,195 @@
+        **description_param(field)
+    )
+
+
+@type_message.register(RepeatField)
+def type_message_repeat(field: RepeatField) -> TypeMessage:
+    return TypeMessage(
+        type=ARRAY,
+        items=type_message(field.items),
+        **description_param(field)
+    )
+
+
+@type_message.register(MapField)
+def type_message_map(field: MapField) -> TypeMessage:
+    return TypeMessage(
+        type='object',
+        additionalProperties=type_message(field.values),
+        **description_param(field)
+    )
+
+
+def reference_string(message: Message) -> str:
+    name = message.__meta__.name
+    if name == 'Empty':
+        return ''
+    return '#/{}/{}'.format(DEFINITIONS, name)
+
+
+def parameter_common(field: Field) -> dict:
+    result = {'name': field.name}
+    f_type = type_message(field)
+    protocol = JSON(type(f_type))
+    result.update(protocol.encode(f_type))
+    return result
+
+
+def parameters_path(method: Method) -> list:
+    fields = [getattr(method.request, f) for f in method.http_path_parameters()]
+    return [{**PATH_PARAMETER, **parameter_common(f)} for f in fields]
+
+
+def parameters_query(method: Method) -> list:
+    fields = [getattr(method.request, f)
+              for f in method.http_field_locations()[HTTPFieldLocation.QUERY]]
+    return [{**QUERY_PARAMETER, **parameter_common(f)} for f in fields]
+
+
+def parameters_body(method: Method) -> list:
+    body_fields = method.http_field_locations()[HTTPFieldLocation.BODY]
+    if not body_fields:
+        return []
+    fields = {f: getattr(method.request, f) for f in body_fields}
+    if fields == method.request.__fields__:
+        param = dict(
+            name=method.request.__meta__.name,
+            schema=TypeMessage(ref=reference_string(method.response))
+        )
+    else:
+        param = dict(
+            name=method.name + '_body',
+            schema=fields_message(fields.values())
+        )
+    return [{**BODY_PARAMETER, **param}]
+
+
+def to_message(param_data: dict) -> ParameterMessage:
+    return ParameterMessage(
+        **{k: param_data[k] for k in field_names(ParameterMessage) if k in param_data}
+    )
+
+
+def parameters_messages(m: Method) -> list:
+    return list(map(to_message, parameters_body(m) + parameters_path(m) + parameters_query(m)))
+
+
+def response_message(method: Method) -> ResponseMessage:
+    return ResponseMessage(
+        description=method.options.get(DESCRIPTION, ''),
+        schema=FieldsMessage(ref=reference_string(method.response))
+    )
+
+
+def method_message(method: Method) -> MethodMessage:
+    return MethodMessage(
+        produces=['application/json'],  # TODO: other formats
+        responses=ResponsesMessage(default=response_message(method)),
+        parameters=parameters_messages(method)
+    )
+
+
+def methods_messages_map(reflect: Reflect) -> dict:
+    result = defaultdict(dict)
+    for k, group in groupby(reflect.methods, key=lambda a: a.http_path):
+        for m in group:
```

Comment:
Again, this function should probably just be in the one function where it is used. It is okay to split up code for reuse and for reducing depth, but it is also very hard to follow the logic if there is no linear flow anywhere.